### PR TITLE
Add automatic API tracking for schemas

### PR DESCRIPTION
**Pull request by JakeWharton**
This adds a `redwoodApiCheck` task tied to the `check` lifecycle task, and a `redwoodApiGenerate` for updating the file.

Refs #37

### CODE REVIEW

File: redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodSchemaPlugin.kt [hunk 1]
```diff
@@ -52,22 +57,54 @@ public class RedwoodSchemaPlugin : Plugin<Project> {
...
+        val apiCheck =
+          project.tasks.register("redwoodApiCheck", RedwoodSchemaApiCheckTask::class.java) {
+            it.group = VERIFICATION_GROUP
+            it.description = "Validate the API tracking file against the latest schema"
+
+            it.toolClasspath.from(toolingConfiguration)
+            it.apiFile.set(apiFile)
+            it.schemaType.set(extension.type)
+            it.classpath.from(classpath, compilation.output.classesDirs)
+            it.dummyOutputFile.set(project.layout.buildDirectory.file("tmp/redwoodApiCheckDummy.txt"))
```
Comment (swankjesse) [inline redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodSchemaPlugin.kt:99]: Comment for this file? Surprises me!

File: redwood-gradle-plugin/src/test/kotlin/app/cash/redwood/gradle/FixtureTest.kt [hunk 2]
```diff
@@ -18,24 +18,92 @@ package app.cash.redwood.gradle
...
+      assertThat(result.output).contains(
+        """
+        |API file redwood-api.xml missing!
+        |
+        |Run 'redwoodApiGenerate' to generate this file.
+        |
+        """.trimMargin(),
+      )
+    }
+    fixtureGradleRunner(fixtureDir, "redwoodApiGenerate").build()
```
Comment (swankjesse) [inline redwood-gradle-plugin/src/test/kotlin/app/cash/redwood/gradle/FixtureTest.kt:103]: Running it multiple times here is nice

File: redwood-gradle-plugin/src/test/kotlin/app/cash/redwood/gradle/FixtureTest.kt [hunk 3]
```diff
@@ -18,24 +18,92 @@ package app.cash.redwood.gradle
...
     assertThat(fixtureDir.resolve("widget/build/generated/redwood").walk().toList()).isNotEmpty()
   }
 
+  @Test fun schemaApiClean() {
+    val fixtureDir = File("src/test/fixture/schema-api-clean")
+    fixtureGradleRunner(fixtureDir, "clean", "check").build().let { result ->
+      val redwoodApiCheck = result.task(":redwoodApiCheck")!!
+      assertThat(redwoodApiCheck).prop(BuildTask::getOutcome).isEqualTo(SUCCESS)
+    }
+    // Ensure the task is properly incremental.
```
Comment (swankjesse) [inline redwood-gradle-plugin/src/test/kotlin/app/cash/redwood/gradle/FixtureTest.kt:51]: Ooooh I want this on the Zipline tests

File: redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodSchemaPlugin.kt [hunk 4]
```diff
@@ -52,22 +57,54 @@ public class RedwoodSchemaPlugin : Plugin<Project> {
...
       it.toolClasspath.from(toolingConfiguration)
       it.outputDir.set(project.layout.buildDirectory.dir("generated/redwood"))
       it.schemaType.set(extension.type)
       it.classpath.from(classpath, compilation.output.classesDirs)
     }
+    project.tasks.named(JAR_TASK_NAME, Jar::class.java).configure {
+      it.from(generateJson)
+    }
+
+    project.afterEvaluate {
```
Comment (swankjesse) [inline redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodSchemaPlugin.kt:74]: Note or TODO on why the naughty afterEvaluate is needed? I assume it’s related to `compilation` not being ready?

File: redwood-tooling-schema/src/main/kotlin/app/cash/redwood/tooling/schema/ApiCommand.kt [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  name = "api",
+  help = "Write schema protocol API to XML, or validate schema compatibility with existing XML",
+) {
+  private val file by option("-f", "--file")
+    .path()
+    .required()
+    .help("XML file")
+
+  private val classpath by option("-cp", "--class-path")
+    .convert { it.split(File.pathSeparator).map(::File) }
```
Comment (swankjesse) [inline redwood-tooling-schema/src/main/kotlin/app/cash/redwood/tooling/schema/ApiCommand.kt]:
```suggestion
        .file()
        .split(File.pathSeparator)
```

File: redwood-tooling-schema/src/main/kotlin/app/cash/redwood/tooling/schema/ApiCommand.kt [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    .required()
+
+  private val mode by option("-m", "--mode")
+    .choice("check" to Check, "generate" to Generate)
```
Comment (swankjesse) [inline redwood-tooling-schema/src/main/kotlin/app/cash/redwood/tooling/schema/ApiCommand.kt:52]: I will rename Zipline’s from `Dump` to `Generate`. That’s nicer.

File: redwood-tooling-schema/src/main/kotlin/app/cash/redwood/tooling/schema/apiModel.kt [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+        }
+        appendLine()
+        append(
+          if (differences.any { it.type == Fatal }) {
+            """
+            The API file cannot be updated automatically due to the presence of fatal
+            differences. Either revert the fatal differences to the schema, or delete the
+            file and run '$fixCommand' to regenerate. Note that regenerating the file
+            means the new schema is fundamentally incompatible with the old one and will
+            produce errors if both the host and guest code are not updated together.
```
Comment (swankjesse) [inline redwood-tooling-schema/src/main/kotlin/app/cash/redwood/tooling/schema/apiModel.kt:98]: Love this

File: redwood-tooling-schema/src/main/kotlin/app/cash/redwood/tooling/schema/apiModel.kt [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+    private val xml = XML {
+      indent = 2
+      autoPolymorphic = true
+    }
+
+    /**
+     * Compute the difference between [this] and [previous] collections by [key].
+     * Returns a triple of matching pairs, added items, and removed items.
+     */
+    private fun <T, K> Collection<T>.diffFrom(
```
Comment (swankjesse) [inline redwood-tooling-schema/src/main/kotlin/app/cash/redwood/tooling/schema/apiModel.kt:235]: Nice Nice Nice

File: redwood-tooling-schema/src/test/kotlin/app/cash/redwood/tooling/schema/ApiTest.kt [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+      | - Widget(tag=2) is missing from the file, but it is a part of the Kotlin schema (fixable)
+      | - Widget(tag=1) type changed from com.example.Widget to com.example.WidgetTypeChange in the Kotlin schema (warning)
+      | - Widget(tag=1) is missing property(tag=3) which is part of the Kotlin schema (fixable)
+      | - Widget(tag=1) is missing event(tag=6) which is part of the Kotlin schema (fixable)
+      | - Widget(tag=1) is missing children(tag=3) which is part of the Kotlin schema (fixable)
+      | - Widget(tag=1) property(tag=1) name changed from first to firstNameChange (fixable)
+      | - Widget(tag=1) event(tag=4) name changed from event to eventNameChange (fixable)
+      | - Widget(tag=1) event(tag=5) nullability changed from true to false (fixable)
+      | - Widget(tag=1) children(tag=1) name changed from left to leftNameChange (fixable)
+      | - Modifier(tag=2) is missing from the file, but it is part of the Kotlin schema (fixable)
```
Comment (swankjesse) [inline redwood-tooling-schema/src/test/kotlin/app/cash/redwood/tooling/schema/ApiTest.kt:182]: These (fixable) tips are great

**Review by swankjesse — APPROVED** [review]
Having the FIR stuff completely independent is so nice.

File: redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodSchemaPlugin.kt (hunk 1 above)
Comment (JakeWharton) [reply redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodSchemaPlugin.kt:99]: There's KDoc on the `dummyOutputFile` property inside the task. I'll move it or duplicate part of it here.

File: redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodSchemaPlugin.kt (hunk 4 above)
Comment (JakeWharton) [reply redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodSchemaPlugin.kt:74]: `afterEvaluate` allows the user's script to run and set the boolean which controls whether these tasks get created or not. The alternative is to always create the tasks and have them no-op if the boolean is set.

File: redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodSchemaPlugin.kt (hunk 1 above)
Comment (JakeWharton) [reply redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodSchemaPlugin.kt:99]: Put a snippet here, too.

File: redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodSchemaPlugin.kt (hunk 4 above)
Comment (JakeWharton) [reply redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodSchemaPlugin.kt:74]: Commented

File: redwood-tooling-schema/src/main/kotlin/app/cash/redwood/tooling/schema/ApiCommand.kt (hunk 5 above)
Comment (JakeWharton) [reply redwood-tooling-schema/src/main/kotlin/app/cash/redwood/tooling/schema/ApiCommand.kt]: Had four of these. Updated all.

**Pull request merged**